drivers/interrupt_controller/intc_gicv3.c

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright 2020 Broadcom
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <assert.h>
+#include <sw_isr_table.h>
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <drivers/interrupt_controller/gic.h>
+
+/* TODO: probe for each implemented rdist base */
+#define GIC_RDIST_BASE  DT_REG_ADDR_BY_IDX(DT_INST(0, arm_gic), 1)
+
+#define GICR_SGIBASE_OFFSET	0x10000
+
+/* Offsets from GICD base or GICR sgi_base */
+#define	GIC_CTRL		0x0
+#define	GIC_TYPER		0x4
+#define	GIC_IIDR		0x8
+#define	GIC_IGROUPR		0x80
+#define	GIC_ISENABLER		0x100
+#define	GIC_ICENABLER		0x180
+#define	GIC_ISPENDR		0x200
+#define	GIC_ICPENDR		0x280
+#define	GIC_ISACTIVER		0x300
+#define	GIC_ICACTIVER		0x380
+#define	GIC_IPRIORITYR		0x400
+#define	GIC_ITARGETSR		0x800
+#define	GIC_ICFGR		0xc00
+#define GIC_IGROUPMODR		0xd00
+#define	GIC_SGIR		0xf00
+
+#define	GICR_CTLR		GIC_RDIST_BASE
+
+#define GICR_WAKER		0x14
+#define GIC_WAKER_PS		0x1
+#define GIC_WAKER_CA		0x2
+
+#define IGROUPR(base, n)	(base + GIC_IGROUPR + (n) * 4)
+#define ISENABLER(base, n)	(base + GIC_ISENABLER + (n) * 4)
+#define ICENABLER(base, n)	(base + GIC_ICENABLER + (n) * 4)
+#define ISPENDR(base, n)	(base + GIC_ISPENDR + (n) * 4)
+#define ICPENDR(base, n)	(base + GIC_ICPENDR + (n) * 4)
+#define IPRIORITYR(base, n)	(base + GIC_IPRIORITYR + n)
+#define ITARGETSR(base, n)	(base + GIC_ITARGETSR + (n) * 4)
+#define ICFGR(base, n)		(base + GIC_ICFGR + (n) * 4)
+#define IGROUPMODR(base, n)	(base + GIC_IGROUPMODR + (n) * 4)
+
+/* Interrupt group definitions */
+#define ENABLE_G0		0
+#define ENABLE_G1NS		1
+#define ENABLE_G1S		2
+#define RWP			31
+
+/*
+ * wait for read write progress
+ * TODO: add timed wait
+ */
+static int gic_wait_rwp(u32_t intid)
+{
+	mem_addr_t base;
+
+	base = (intid < GIC_SPI_INT_BASE) ? GICR_CTLR : GICD_CTLR;
+
+	while (sys_read32(base) & BIT(RWP))
+		;
+
+	return 0;
+}
+
+void arm_gic_irq_set_priority(unsigned int intid,
+			      unsigned int prio, u32_t flags)
+{
+	u32_t bit = intid & (NUM_INTR_PER_REG - 1);
+	u32_t mask = BIT(bit);
+	u32_t shift;
+	u32_t idx = intid / NUM_INTR_PER_REG;
+	u32_t val;
+	mem_addr_t base;
+
+	base = (intid < GIC_SPI_INT_BASE) ?
+	       (GIC_RDIST_BASE + GICR_SGIBASE_OFFSET)
+	       : GIC_DIST_BASE;
```

Comment:
This implementation assumes that there is only one redistributor, and will effectively configure incorrect redistributor in the multi-core systems (normally, there is one redistributor per core).

Since we do not support multi-core systems at this time, we might as well just use the distributor base here, whose index 0 registers map to the local core redistributor register for the systems with less than 8 cores (e.g. `GICD_IPRIORITYR0` is local core `GICR_IPRIORITYR0`).

Reply:
> This implementation assumes that there is only one redistributor

Yes.  Multi core redistributor is part of TODO list.

> Since we do not support multi-core systems at this time, we might as well just use the distributor base here, whose index 0 registers map to the local core redistributor register for the systems with less than 8 cores (e.g. GICD_IPRIORITYR0 is local core GICR_IPRIORITYR0).

What you have have described is V2 legacy mode support which was actually a stopgap to transition to V3 and now it is deprecated. Like I have mentioned this implementation do not support legacy mode. Though it can also be supported in this driver ( effectively supporting GICv2 as well) IMHO there will be no users of it.  However system interface is a must for GICv3 compatible hardware and architecturally complaint software.

Reply:
I cannot find any references to the distributor register index 0 banking being a legacy feature in the GICv3 specification. The only constraint is that `Processor_Number < 8`. Am I missing something here?

IPRIORITYR0, for instance:
![image](https://user-images.githubusercontent.com/11345418/80870011-b07f9100-8cde-11ea-99a4-3b2a50d5e45b.png)

Reply:
Never mind, found it:
![image](https://user-images.githubusercontent.com/11345418/80870152-9db98c00-8cdf-11ea-9af8-d931b2843c96.png)